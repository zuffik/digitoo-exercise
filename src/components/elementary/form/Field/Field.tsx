import * as React from 'react';
import {Field as FormikField} from 'formik';
import classNames from 'classnames';
import styles from './Field.module.sass';

interface Props extends Omit<JSX.IntrinsicElements['input'], 'placeholder'> {
  error?: string;
  touched?: boolean;
  label: string;
  disableLabel?: boolean;
}

export const Field: React.FC<Props> = ({error, touched, label, ...props}: Props) => {
  return (
    <div className="form-group">
      <label className={styles.field}>
        {!props.disableLabel && <span className="mb-2 d-block">{label}</span>}
        <FormikField
          {...props}
            data-testid="field-input"
          placeholder={label}
          className={classNames(props.className, 'form-control', {
            'is-invalid': touched && error,
            'is-valid': touched && !error,
          })}
        />
        <small className="text-danger form-text" data-testid="field-helper-text">{(touched && error) || ' '}</small>
      </label>
    </div>
  );
};
