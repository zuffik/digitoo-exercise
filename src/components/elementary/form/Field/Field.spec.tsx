import * as React from 'react';
import {render} from '@testing-library/react'
import {Field} from "./Field";
import {Formik} from "formik";

describe('<Field/>', () => {
    it('should render', () => {
        const {asFragment, getByTestId, queryAllByText} = render(
            <Formik initialValues={{}} onSubmit={jest.fn()}>
                <Field label="Label"/>
            </Formik>
        );
        expect(queryAllByText('Label')).toHaveLength(1);
        expect(getByTestId('field-input').className).not.toMatch(/(is-valid|is-invalid)/);
        expect(getByTestId('field-helper-text')).toHaveTextContent('');
        expect(asFragment()).toMatchSnapshot();
    });
    it('should render with error', () => {
        const {asFragment, getByTestId, queryAllByText} = render(
            <Formik initialValues={{}} onSubmit={jest.fn()}>
                <Field label="Label" error="Error" touched/>
            </Formik>
        );
        expect(queryAllByText('Label')).toHaveLength(1);
        expect(getByTestId('field-input').className).toMatch(/is-invalid/);
        expect(getByTestId('field-helper-text')).toHaveTextContent('Error');
        expect(asFragment()).toMatchSnapshot();
    });
    it('should render with success', () => {
        const {asFragment, getByTestId, queryAllByText} = render(
            <Formik initialValues={{}} onSubmit={jest.fn()}>
                <Field label="Label" touched/>
            </Formik>
        );
        expect(queryAllByText('Label')).toHaveLength(1);
        expect(getByTestId('field-input').className).toMatch(/is-valid/);
        expect(getByTestId('field-helper-text')).toHaveTextContent('');
        expect(asFragment()).toMatchSnapshot();
    });
    it('should render without label', () => {
        const {asFragment, getByTestId, queryAllByText} = render(
            <Formik initialValues={{}} onSubmit={jest.fn()}>
                <Field label="Label" disableLabel/>
            </Formik>
        );
        expect(queryAllByText('Label')).toHaveLength(0);
        expect(getByTestId('field-input').className).not.toMatch(/(is-valid|is-invalid)/);
        expect(getByTestId('field-helper-text')).toHaveTextContent('');
        expect(asFragment()).toMatchSnapshot();
    });
});
