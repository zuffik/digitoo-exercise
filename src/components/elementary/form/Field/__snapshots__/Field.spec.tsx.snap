// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Field/> should render 1`] = `
<DocumentFragment>
  <div
    class="form-group"
  >
    <label
      class="field"
    >
      <span
        class="mb-2 d-block"
      >
        Label
      </span>
      <input
        class="form-control"
        data-testid="field-input"
        placeholder="Label"
        value="[object Object]"
      />
      <small
        class="text-danger form-text"
        data-testid="field-helper-text"
      >
         
      </small>
    </label>
  </div>
</DocumentFragment>
`;

exports[`<Field/> should render with error 1`] = `
<DocumentFragment>
  <div
    class="form-group"
  >
    <label
      class="field"
    >
      <span
        class="mb-2 d-block"
      >
        Label
      </span>
      <input
        class="form-control is-invalid"
        data-testid="field-input"
        placeholder="Label"
        value="[object Object]"
      />
      <small
        class="text-danger form-text"
        data-testid="field-helper-text"
      >
        Error
      </small>
    </label>
  </div>
</DocumentFragment>
`;

exports[`<Field/> should render with success 1`] = `
<DocumentFragment>
  <div
    class="form-group"
  >
    <label
      class="field"
    >
      <span
        class="mb-2 d-block"
      >
        Label
      </span>
      <input
        class="form-control is-valid"
        data-testid="field-input"
        placeholder="Label"
        value="[object Object]"
      />
      <small
        class="text-danger form-text"
        data-testid="field-helper-text"
      >
         
      </small>
    </label>
  </div>
</DocumentFragment>
`;

exports[`<Field/> should render without label 1`] = `
<DocumentFragment>
  <div
    class="form-group"
  >
    <label
      class="field"
    >
      <input
        class="form-control"
        data-testid="field-input"
        placeholder="Label"
        value="[object Object]"
      />
      <small
        class="text-danger form-text"
        data-testid="field-helper-text"
      >
         
      </small>
    </label>
  </div>
</DocumentFragment>
`;
